source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,235 @@
+  public:
+    LeafNode(absl::string_view key, Value&& value) : key_(key), value_(std::move(value)) {}
+    Value find(const absl::string_view& key) override {
+      // String comparison unnecessarily checks size equality first, we can skip
+      // to memcmp here because we already know the sizes are equal.
+      // Since this is a super-hot path we don't even ASSERT here, to avoid adding
+      // slowdown in debug builds.
+      if (memcmp(key.data(), key_.data(), key.size())) {
+        return {};
+      }
+      return value_;
+    }
+
+  private:
+    const std::string key_;
+    const Value value_;
+  };
+
+  class BranchNode : public Node {
+  public:
+    BranchNode(size_t index, uint8_t min, std::vector<std::unique_ptr<Node>>&& branches)
+        : index_(index), min_(min), branches_(std::move(branches)) {}
+    Value find(const absl::string_view& key) override {
+      uint8_t k = static_cast<uint8_t>(key[index_]);
+      // Possible optimization was tried here, populating empty nodes with
+      // a function that returns {} to reduce branching vs checking for null
+      // nodes. Checking for null nodes benchmarked faster.
+      if (k < min_ || k >= min_ + branches_.size() || branches_[k - min_] == nullptr) {
+        return {};
+      }
+      return branches_[k - min_]->find(key);
+    }
+
+  private:
+    const size_t index_;
+    const uint8_t min_;
+    // Possible optimization was tried here, using std::array<std::unique_ptr<Node>, 256>
+    // rather than a smaller-range vector with bounds, to keep locality and reduce
+    // comparisons. It didn't help.
+    const std::vector<std::unique_ptr<Node>> branches_;
+  };
+
+public:
+  // The caller owns the string-views during `compile`. Ownership of the passed in
+  // Values is transferred to the CompiledStringMap.
+  using KV = std::pair<absl::string_view, Value>;
+  /**
+   * Returns the value with a matching key, or the default value
+   * (typically nullptr) if the key was not present.
+   * @param key the key to look up.
+   */
+  Value find(absl::string_view key) const {
+    const size_t key_size = key.size();
+    if (key_size >= table_.size() || table_[key_size] == nullptr) {
+      return {};
+    }
+    return table_[key_size]->find(key);
+  };
+  /**
+   * Construct the lookup table. This can be a somewhat slow multi-pass
+   * operation if the input table is large.
+   * @param initial a vector of key->value pairs. This is taken by value because
+   *                we're going to modify it. If the caller still wants the original
+   *                then it can be copied in, if not it can be moved in.
+   *                Note that the keys are string_views - the base string data must
+   *                exist for the duration of compile(). The leaf nodes take copies
+   *                of the key strings, so the string_views can be invalidated once
+   *                compile has completed.
+   */
+  void compile(std::vector<KV> initial) {
```

Comment:
I wonder if there needs to be a dedup of the keys?
I think that in the original code it would've been overwritten, but in this data-structure case, I'm not so sure what will be the implications.